lib/Sema/CSDiagnostics.cpp

```diff
@@ -88,19 +90,54 @@ Type RequirementFailure::getOwnerType() const {
 }
 
 const GenericContext *RequirementFailure::getGenericContext() const {
+  // For conditional requirements, this is easy because all
+  // required information is contained in the conformance
+  // reference.
+  if (isConditional()) {
+    auto *conformance = getConformanceRef().getConcrete();
+    auto *DC = conformance->getDeclContext();
```

Comment:
We should think of a better way to express this. I think I've see this appear a few times in this file.

Reply:
Actually I realized that I can simply do `getRequirementDC()` in this case, no need to fetch conformance explicitly, thanks!

Reply:
Hm, after thinking about this - it genericContext() should be related to the affect decl here and not the requirement, so this if shouldn't even be needed...

Reply:
Fixed!

Reply:
That also improved diagnostic for:
```swift
extension BinaryInteger {
  var factorial: Self {
    return self <= 1
                ? 1
                : (2...self).reduce(1, *)
  }
}
```

To:
```
error: referencing instance method 'reduce' on 'ClosedRange' requires that 'Self.Stride' conform to 'SignedInteger'
                  : (2...self).reduce(1, *)
                               ^
note: requirement from conditional conformance of 'ClosedRange<Self>' to 'Sequence'
extension ClosedRange : Sequence where Bound : Strideable, Bound.Stride : SignedInteger {
          ^
```